src/git/repo.rs

```diff
@@ -1446,6 +1672,98 @@ impl<'repo> Branch<'repo> {
     }
 }
 
+/// A Git file status indicator.
+/// See https://git-scm.com/docs/git-status#_short_format
+#[allow(missing_docs)]
+#[derive(Debug, PartialEq, Eq)]
+pub enum FileStatus {
+    Unmodified,
+    Modified,
+    Added,
+    Deleted,
+    Renamed,
+    Copied,
+    Unmerged,
+    Untracked,
+    Ignored,
+}
+
+/// The status of a file in the repo.
+#[derive(Debug, PartialEq, Eq)]
+pub struct StatusEntry {
+    /// The status of the file in the index.
+    pub index_status: FileStatus,
+    /// The status of the file in the working tree.
+    pub work_tree_status: FileStatus,
+    /// The file path.
+    pub path: String,
+    /// The original path of the file (for renamed files).
+    pub orig_path: Option<String>,
+}
+
+impl StatusEntry {
+    /// Returns the paths associated with the status entry.
+    pub fn paths(&self) -> Vec<PathBuf> {
+        let mut result = vec![PathBuf::from(self.path.clone())];
+        if let Some(orig_path) = &self.orig_path {
+            result.push(PathBuf::from(orig_path.clone()));
+        }
+        result
+    }
+}
+
+fn parse_file_status(status: char) -> eyre::Result<FileStatus> {
+    let status = match status {
+        ' ' => FileStatus::Unmodified,
+        'M' => FileStatus::Modified,
+        'A' => FileStatus::Added,
+        'D' => FileStatus::Deleted,
+        'R' => FileStatus::Renamed,
+        'C' => FileStatus::Copied,
+        'U' => FileStatus::Unmerged,
+        '?' => FileStatus::Untracked,
+        '!' => FileStatus::Ignored,
+        _ => eyre::bail!("invalid status indicator: {}", status),
+    };
+    Ok(status)
+}
+
+impl FromStr for StatusEntry {
```

Comment:
File paths may be any sequence of non-null bytes, not just UTF-8-encoded strings. Let's parse those properly. (There are a few cases in the rest of the program where we operate on strings instead of `PathBuf`s, but those are due to libgit2 limitations, and they could be fixed in the future.)

To operate on bytes, you'll want to use the `&[u8]` type. I think a lot of iterator methods also end up working on byte sequences, such as `split`. You can also use `regex::bytes` if necessary.

I saw this comment in the documentation, which should make parsing the output from `git status` easier:

> There is also an alternate -z format recommended for machine parsing. In that format, the status field is the same, but some other things change. First, the -> is omitted from rename entries and the field order is reversed (e.g from -> to becomes to from). Second, a NUL (ASCII 0) follows each filename, replacing space as a field separator and the terminating newline (but a space still separates the status field from the first filename). Third, filenames containing special characters are not specially formatted; no quoting or backslash-escaping is performed.